### PR TITLE
Fix form validation message

### PR DESCRIPTION
**Pull request by chen-robert**
Fixes https://github.com/publiclab/plots2/issues/6303

I noticed that `user.save` populated many of the relevant errors. This was not being called when the recaptcha failed because the if statement short circuited the method call. 

```ruby
if ((@spamaway&.valid?) || recaptcha) && @user.save
```

To make this code more declarative, I refactored the result of `@user.save` into a new variable `saved`. This way, the errors would be populated every time. 

In addition, I added code to manually clean up the errors returned from the `@user.save`. In particular, the `is invalid` errors aren't very informative. I added a check to remove these errors if there was already an error associated with that field. 

Finally, I added an integration test for the errors returned from the `/register` endpoint. 

@gauravano @SidharthBansal @IshaGupta18 @ananyaarun @pydevsg @sashadev-sky @cesswairimu Could you take a look please? Thanks!

### CODE REVIEW

**Comment by SidharthBansal** [comment]
I have to change it to hard 
Well done
Lgtm :tada:
Waiting for passing test with proper sign up details. I think it already exists. So pls give me it's li nk
Thanks

**Comment by chen-robert** [comment]
@SidharthBansal The registration with proper signup details test can be found here. 

https://github.com/publiclab/plots2/blob/master/test/integration/login_flow_test.rb#L34-L55

Thanks!

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -14,7 +14,9 @@ def create
     using_recaptcha = !params[:spamaway] && Rails.env == "production"
     recaptcha = verify_recaptcha(model: @user) if using_recaptcha
     @spamaway = Spamaway.new(spamaway_params) unless using_recaptcha
-    if ((@spamaway&.valid?) || recaptcha) && @user.save
+
+    saved = @user.save
+    if saved && ((@spamaway&.valid?) || recaptcha)
```
[the remark this answers is not in the document]
Comment (chen-robert) [reply app/controllers/users_controller.rb]: Originally when I did not refactor `@user.save` out into a new variable, the ordering of the two statements mattered because the if statement would shortcircuit if recaptcha failed. This is no longer an issue now, and the ordering of the statement doesn't matter.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -47,6 +49,17 @@ def create
       end
       # send all errors to the page so the user can try again
       @action = "create"
+      email_errs = @user.errors.messages[:email]
+      username_errs = @user.errors.messages[:username]
+
+      if email_errs.size > 1
```
[the remark this answers is not in the document]
Comment (chen-robert) [reply app/controllers/users_controller.rb]: There are multiple errors for the username, such as `Email should look like an email address` which is redundant from `Email is invalid`. I check if the size of `email_errs` is greater than one, which implies that additional errors other than just `is invalid` exists. If this is the case, we can safely remove the `Email is invalid` error which provides no new information.

File: app/controllers/users_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (chen-robert) [reply app/controllers/users_controller.rb]: I can change this back if needed

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -47,6 +49,17 @@ def create
...
+        email_errs.delete("is invalid")
+      end
+
+      if username_errs.size > 1
```
[the remark this answers is not in the document]
Comment (chen-robert) [reply app/controllers/users_controller.rb]: Same as the email issue

File: app/controllers/users_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (SidharthBansal) [reply app/controllers/users_controller.rb]: If this change is redundant to this pr then please remove this change and edit the issue body.
Thansk

File: app/controllers/users_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (SidharthBansal) [reply app/controllers/users_controller.rb]: Problem statement: Email is too short, Email is too long, ....
should be present according to the email entered
Your Proposed Soln: Email Invalid if more than 1 error
User is still facing problem: What I did wrong?(User is thinking)

[the remark this answers is not in the document]
Comment (SidharthBansal) [reply app/controllers/users_controller.rb]: Btw `email_errs` like short forms are discouraged in production code

[the remark this answers is not in the document]
Comment (SidharthBansal) [reply app/controllers/users_controller.rb]: Example: Username is 123$
Intended behaviour
Error 1: Too short
Error 2: Username should only contain A-Z, a-z, ., @,_.

[the remark this answers is not in the document]
Comment (SidharthBansal) [reply app/controllers/users_controller.rb]: Suggestion: Please confirm this error still exists on the localhost and main prod website. It has been 2 years I guess. May be some code fixed it

[the remark this answers is not in the document]
Comment (chen-robert) [reply app/controllers/users_controller.rb]: If you try signing up with a blank username, it gives the following errors
image
The `Username is invalid` error is redundant and can be safely removed, which I have implemented.

[the remark this answers is not in the document]
Comment (SidharthBansal) [reply app/controllers/users_controller.rb]: Tell me why will Username has already been taken come?

**Comment by chen-robert** [comment]
Could you rephrase that question? Sorry I don't understand what you're asking

**Comment by SidharthBansal** [comment]
Tell me what we need in this issue?
I think you are not able to think about the problem statement correctly!

File: app/controllers/users_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (SidharthBansal) [reply app/controllers/users_controller.rb]: I think you forget to remove these redundant changes

**Comment by chen-robert** [comment]
Currently the error messages for emails looks like this. 

image

I want it to look like 

image

Is this not intended?

**Comment by SidharthBansal** [comment]
No

**Comment by SidharthBansal** [comment]
Think about your mistake while I am typing

**Comment by chen-robert** [comment]
Should I leave in the `email is invalid` error then?

**Comment by chen-robert** [comment]
* username is invalid

**Comment by SidharthBansal** [comment]
Currently, the error messages for emails look like this:
Input: ``(blank)

image

I want it to look like 

`Username should not be blank`

OR

`Username is too short(minimum 3 characters)`

**Comment by chen-robert** [comment]
Just to clarify, do you mean only one username error should display at a time?

**Comment by chen-robert** [comment]
For example, if the submitted username was `^` (not a valid username by length and character type), should I display both `Username should use only letters, numbers, spaces, and .-_@+ please.` and `Username is too short (minimum is 3 characters)` or just one of them.

**Comment by SidharthBansal** [comment]
We want to do it for both types of error. I am just discussing username first. You can do same thing for the other attribute.
Thanks

**Comment by SidharthBansal** [comment]
> For example, if the submitted username was ^ (not a valid username by length and character type), should I display both Username should use only letters, numbers, spaces, and .-_@+ please. and Username is too short (minimum is 3 characters) or just one of them.

Both

**Comment by chen-robert** [comment]
So if I submitted a blank username, it should show only `Username is too short (minimum is 3 characters)` ?

**Comment by SidharthBansal** [comment]
Bonus: Create a fto for changing `Username should use only letters, numbers, spaces, and .-_@+ please.` -> `Please use letters, numbers, spaces and -_@+ ONLY in the usernames`

This is not related to the task we are trying to complete. It is a side task which may give you some extra points.
Extra bonus: Use similar tone for other errors and create a lot of ftos using FTO template. 

Tip: Get your first 2-3 ftos approved by me before moving forward and creating more FTOs

**Comment by chen-robert** [comment]
Okay, I'll do that, thanks for the idea!

**Comment by SidharthBansal** [comment]
@chen-robert if you enter no username it should either display
`Please don't leave username blank`
OR
`Username is too short (minimum is 3 characters)`

**Comment by SidharthBansal** [comment]
Anything else you wanna know or shall I sleep?
It;s 2pm here

**Comment by chen-robert** [comment]
I think I understand it now, thanks for all the help!

**Comment by SidharthBansal** [comment]
Good night. See you soon!
Login issues are mostly reviewed by me. I am very busy sometimes, so please
do the above ftos if I can't help you and you are free. This way you will
be able to utilize your time and get points

**Comment by chen-robert** [comment]
I editted the `user.rb` model to add a custom validation error for blank usernames. I also changed the regex which was breaking on usernames of length 1. 

The error now looks as such.

image

**Comment by SidharthBansal** [comment]
Great!
Write test for all the sequences you think should be tested. If something
breaks then do changes in the code accordingly
I can review it after 9 hours. May be after a day or two. But will try to
do it asap. Goodnight 😴

**Comment by chen-robert** [comment]
Thanks! Good night :sleeping:

**Comment by SidharthBansal** [comment]
Can you please see other cases too like username is 'xzd@' so only one error will come. Username can't contain @_-
Similarly for 'z@' two errors should come  min length and special correct error.
Please check more test cases

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -14,7 +14,9 @@ def create
     using_recaptcha = !params[:spamaway] && Rails.env == "production"
     recaptcha = verify_recaptcha(model: @user) if using_recaptcha
     @spamaway = Spamaway.new(spamaway_params) unless using_recaptcha
-    if ((@spamaway&.valid?) || recaptcha) && @user.save
+
+    saved = @user.save
+    if ((@spamaway&.valid?) || recaptcha) && saved
```
Comment (SidharthBansal) [inline app/controllers/users_controller.rb]: If redundant remove it

**Comment by SidharthBansal** [comment]
Add screenshots too

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (chen-robert) [reply app/controllers/users_controller.rb]: The refactoring is important because it allows `@user.save` to be called everytime, populating the errors object.

File: app/controllers/users_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (chen-robert) [reply app/controllers/users_controller.rb]: This was what shows up on https://publiclab.org/ where the blank username is already registered.

**Comment by chen-robert** [comment]
Here is a screenshot of what the errors look like 

image

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (chen-robert) [reply app/controllers/users_controller.rb]: I explained this in more detail here, https://github.com/publiclab/plots2/pull/6932#issue-350216461

**Comment by SidharthBansal** [comment]
Can you do exhaustive search and provide me all test cases and all
screenshots of all cases?
If present don't write it again
Comments are missing

**Comment by chen-robert** [comment]
There are a total of 12 tests of the `/register` endpoint. 
image

Two already existed, one in `test/integration/login_flow_test.rb`

image

And the other in `test/integration/routes_test.rb`

image


The other 10 were added by me. 

image
image
image
image
image
image
image
image
image
image

**Comment by SidharthBansal** [comment]
I asked for localhost screenshots. Full screens pls.

**Comment by chen-robert** [comment]
@SidharthBansal 
image

**Comment by chen-robert** [comment]
You can find the list of all the tests I wrote in `test/integration/signup_flow_test.rb`.

**Comment by SidharthBansal** [comment]
Do all your work. At last send me corrected screenshots with tests for
exhaustive search. Take some time
Thanks

**Comment by SidharthBansal** [comment]
I can see what you did. :-)
Still things are missing

**Comment by chen-robert** [comment]
Just to clarify, do you want me to upload a screenshot for each of the tests that I already wrote in `test/integration/signup_flow_test.rb`?

**Comment by SidharthBansal** [comment]
Screenshot of working of the flow written in the test on the localhost.

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (SidharthBansal) [reply app/controllers/users_controller.rb]: Can we shift this to another PR please

**Comment by chen-robert** [comment]
So for each of the 14 tests I wrote, I should have one before and one after for a total of 28 screenshots?

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -47,6 +49,17 @@ def create
       end
       # send all errors to the page so the user can try again
       @action = "create"
+      email_errors = @user.errors.messages[:email]
```
Comment (SidharthBansal) [inline app/controllers/users_controller.rb]: can we shift `is invalid` to other PR

Comment (chen-robert) [reply app/controllers/users_controller.rb]: Okay, I will update remove it from this PR

**Comment by SidharthBansal** [comment]
> So for each of the 14 tests I wrote, I should have one before and one after for a total of 28 screenshots?

Correct

**Comment by chen-robert** [comment]
Okay, I understand now. Thanks!

**Comment by SidharthBansal** [comment]
Also, please remember to get 5pts for username and 5pts for email errors.
Ping me to upload one more sign up validation issue so that you can get more points.
Thanks

**Comment by chen-robert** [comment]
@SidharthBansal There is actually a rather tricky bug that took me a while to investigate. 

For some reason, the `Username should use only letters, numbers, spaces, and .-_@+ please.` was appearing even with the one letter username `a`. We would expect that the one letter username `a` would only generate a `Username is too short (minimum is 3 characters)` error. 

After googling this error, I found [the documentation where](https://www.rubydoc.info/gems/authlogic/2.0.1/Authlogic/I18n) it appears to be added as part of the `authlogic` gem. 

Looking back at user model in `app/models/user.rb`, we do indeed use authlogic by calling `acts_as_authentic`. 

https://github.com/publiclab/plots2/blob/d65980307b8a42ce19945485083d340406e81289/app/models/user.rb#L33-L35

However checking the [Github repository](https://github.com/binarylogic/authlogic), the `regex.rb` file doesn't exist! Digging into the PR history, I found that the regex was dropped in [this PR](https://github.com/binarylogic/authlogic/pull/640). 

Looking at the [actual regex expression referenced](https://github.com/binarylogic/authlogic/pull/640/files#diff-dd46fd19af66b21da6f2671cdc0b2001R41) in the PR, `/\A[a-zA-Z0-9_][a-zA-Z0-9\.+\-_@ ]+\z/`, we find that it indeed fails for usernames of length one. 

You can test this yourself on [Rubular, a ruby regex tester](https://rubular.com/) if you want as well. 

image

image

One possible solution would be to update the `authlogic` gem to version 5. In the PR, it gives instructions for updating to version 5 where the validations will be completely dropped.
 
Note that the version currently installed in `plots2` is 4.4.2. 

```bash
 $ bundle list | grep authlogic
  * authlogic (4.4.2)
```

However, this solution may create unexpected side effects related to upgrading the `authlogic` version. 

Another option would be to manually add checks for the `Username should use only letters, numbers, spaces, and .-_@+ please.` error and remove it when it fails for usernames of length one. 

Finally, we could also simply disable the username field validation as described [in this Stackoverflow article](https://stackoverflow.com/a/26995019/10178580). Again, this solution could have unintended side effects by completely disabling the authlogic username validation. 

```ruby
acts_as_authentic do |c|
  c.validate_login_field = false
end
```

**TLDR**
The `authlogic` gem which is used for authentication has an error in its regex. There are three possible solutions. 

1. Upgrade authlogic version
2. Manually check for authlogic's error and remove it 
3. Disable authlogic username validation

cc: @jywarren @gauravano @IshaGupta18 @ananyaarun @sashadev-sky @cesswairimu

**Comment by SidharthBansal** [comment]
This is the reason I divided it into 2 hard bugs. :-|

**Comment by SidharthBansal** [comment]
Can we focus on username in this PR?
Once we are done with it, we will move to emails.
Thanks

**Comment by jywarren** [comment]
Hello all, thanks for this very long discussion, and for your patience on all sides. I know it can be really busy in GCI and everyone's got a lot of different tasks going at once. ❤️ 

Just catching up, it seems like @chen-robert has identified an edge case where a one-character username triggers the wrong validation? 

One thing I will suggest is that this may not conflict with the narrow definition of the original issue; most of the issue can be solved in this PR without needing to /also/ solve the one-character username issue. I suggest we split that off and think about it later - an Authlogic upgrade is good to consider in any case (#6274 for example) and perhaps we can have that be an issue too. 

It sounds like (i may not fully grasp) the core issue in the PR is solved here, is that correct? If so, we should aim to ensure the tests correctly demonstrate this and the single-character edge case seems a) not very likely and b) still doesn't allow people to make a one-character username, which is a good thing. 

Does this seem merge-able pending test resolution? If so, let's split off other issues and address things in focused chunks. Remember: [modularity is great for collaboration](https://publiclab.org/notes/warren/01-18-2018/software-outreach-modularity-is-great-for-collaboration)! It makes both code writing and reviewing easier!

Thanks everybody!

**Comment by chen-robert** [comment]
@jywarren I believe I have a solution for the broken authlogic regex. I've monkeypatched the regex authlogic uses in [this commit](https://github.com/publiclab/plots2/pull/6932/commits/a83e2f64a64aa559cf13bf18b902b2ac0fc619c3). 

```ruby
Authlogic::Regex::LOGIN = /\A[A-Za-z\d_\-]*\z/
```

The username issue (including the one character username bug) should be completely fixed now. The email validation should work as well. I've also written tests to validate both behaviors in `test/integration/signup_flow_test.rb`. 

Thanks for all the patience @SidharthBansal, I appreciate your time and help as we work through this rather difficult issue :tada:

**Comment by SidharthBansal** [comment]
As suggested by us, kindly break this into username pr and email pr.
Thanks

**Comment by chen-robert** [comment]
I will close this PR and break it in two separate PRs as @SidharthBansal suggested. The username validation PR can be found here: https://github.com/publiclab/plots2/pull/6949. 

Thanks for all the help, I really appreciate everyone's time and patience ❤️

**Comment by SidharthBansal** [comment]
For each function please create separate prs. Please add comments too!
